### PR TITLE
ddl: add foreign key check when modify column 

### LINKED ISSUE #37821: modify column should check foreign key constrain
## Dev Task

Modify column should check foreign key constrain.

This is a sub task of https://github.com/pingcap/tidb/issues/36982

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve?

Issue Number: close #37821

Problem Summary: add foreign key check when modify column 

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/foreign_key.go [hunk 1]
```diff
@@ -281,6 +281,68 @@ func checkTableForeignKey(referTblInfo, tblInfo *model.TableInfo, fkInfo *model.
...
+				if referCol == nil {
+					continue
+				}
+				if newCol.GetType() != referCol.GetType() {
+					return dbterror.ErrFKIncompatibleColumns.GenWithStackByArgs(originalCol.Name, fkInfo.RefCols[i], fkInfo.Name)
+				}
+				if newCol.GetFlen() < referCol.GetFlen() || newCol.GetFlen() < originalCol.GetFlen() ||
+					(newCol.GetType() == mysql.TypeNewDecimal && (newCol.GetFlen() != originalCol.GetFlen() || newCol.GetDecimal() != originalCol.GetDecimal())) {
+					return dbterror.ErrForeignKeyColumnCannotChange.GenWithStackByArgs(originalCol.Name, fkInfo.Name)
+				}
```
Comment (lcwangchao) [inline ddl/foreign_key.go:312]: Some codes are the same with the below codes. Can we reuse it ?

**Review by lcwangchao — APPROVED** [review]

File: ddl/foreign_key.go (hunk 1 above)
Comment (crazycs520) [reply ddl/foreign_key.go:312]: It's not exactly the same, then hard to reuse it.

File: ddl/foreign_key_test.go [hunk 2]
```diff
@@ -766,6 +766,40 @@ func TestCreateTableWithForeignKeyError(t *testing.T) {
 	}
 }
 
+func TestModifyColumnWithForeignKey(t *testing.T) {
+	store, _ := testkit.CreateMockStoreAndDomain(t)
```
Comment (hawkingrei) [inline ddl/foreign_key_test.go:770]: ```store := testkit.CreateMockStore(t)```

Comment (crazycs520) [reply ddl/foreign_key_test.go:770]: done.

**Pull request merged**